lib/Event/Traits/DisableEventTrait.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+/**
+ * Pimcore
+ *
+ * This source file is available under two different licenses:
+ * - GNU General Public License version 3 (GPLv3)
+ * - Pimcore Commercial License (PCL)
+ * Full copyright and license information is available in
+ * LICENSE.md which is distributed with this source code.
+ *
+ * @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ * @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Event\Traits;
+
+use Symfony\Contracts\EventDispatcher\Event;
+
+trait DisableEventTrait
+{
+
+    /**
+     * @var array
+     */
+    protected static $disabledEvents = [];
+
+    /**
+     * Sets the list of disabled events
+     *
+     * @param array $events
+     * @return void
+     */
+    public static function disableEvents(array $events = []): void
+    {
+        self::$disabledEvents = $events;
+    }
+
+    /**
+     * Checks if an event is not currently disabled
+     *
+     * @param string $eventName
+     * @return bool
+     */
+    private function isAllowedEvent(string $eventName): bool
+    {
+        return !in_array($eventName, self::$disabledEvents);
+    }
+
+    /**
+     * Dispatches an event, after checking if the event is not previously disabled
+     *
+     * @param Event $event
+     * @param string $name
+     * @return void
+     */
+    protected function dispatchEvent(Event $event, string $name): void
+    {
+        if ($this->isAllowedEvent($name)) {
+            \Pimcore::getEventDispatcher()->dispatch($event, $name);
+        }
+    }
```

Comment:
```suggestion
    protected function dispatchEvent(object $event, string $eventName = null): ?object
    {
        $eventName = $eventName ?? \get_class($event);
        if (!isset($this->activeDispatchingEvents[$eventName])) {
            $this->activeDispatchingEvents[$eventName] = true;
            \Pimcore::getEventDispatcher()->dispatch($event, $eventName);
            unset($this->activeDispatchingEvents[$eventName]);
        }
    }
```
I'd use the same method signature as `\Symfony\Contracts\EventDispatcher\EventDispatcherInterface::dispatch

Reply:
i've tested it, nice approach, works simply fine for the purpose. 

The main difference is that `activeDispatchingEvents` approach is automatical (like by default in the system), while `disableEvents` is more manual and up to the developer:
- the `activeDispatchingEvents` stops the same event that was called by the parent event, but doesn't have control over stopping/preventing any other events from firing. In other words, it won't help to stop PRE_UPDATE in a recursive POST_UPDATE. With `disableEvents`, it would allow to manually decide to stop any other ones before that
-  `disableEvents` could be used in other circumstances beside from preventing recursion.
- if for any reason, there is some recursion intended (likely not, but who knows), that `disableEvents` could work along an exit condition.

Not sure if these could be some points to consider, but the recommendation was to follow the `Version::disable()` approach, that's why it looks like that (like `disableEvents` not used anywhere in the PR).

Reply:
As discussed, we only want to block recursion for the same object at the moment (as this is a bug). 
Can you please update the PR accordingly? Thanks a lot 😊